#' Combines sites and mutations for sites and flanks from each bin in prepared_sites
#'
#' @param maf Data frame of mutations prepared by get_mut_trinuc_strand. Contains all if not cofactors are included
#' @param prepared_sites List of coordinates and quadnucleotide contexts generated by prepare_bins_of_sites
#'
#' @return Data frame containing mutation counts for each quadnucleotide context + indel from each bin and sites + up/down flanks. Also includes trinucleotide position counts and average megabase mutation rate by bin
.maf_to_dfr = function(maf, prepared_sites) {

	if (is.null(prepared_sites[[1]])) {
		return(NULL)
	}

	gr_maf = GenomicRanges::GRanges(maf$chr, IRanges::IRanges(maf$start, maf$end))
	GenomicRanges::mcols(gr_maf)[,"trinuc"] = maf$mut_trinuc
	GenomicRanges::mcols(gr_maf)[,"ID"] = paste(maf$chr, maf$start, maf$end, maf$ref, maf$alt, maf$patient_id, sep=":")

	full_dfr = NULL

	for (i in 1:length(prepared_sites)) {

		gr_sites = prepared_sites[[i]][[1]]
		gr_sites_upstream = prepared_sites[[i]][[2]]
		gr_sites_downstream = prepared_sites[[i]][[3]]

		trinuc_sites = prepared_sites[[i]][[4]]
		trinuc_upstream = prepared_sites[[i]][[5]]
		trinuc_downstream = prepared_sites[[i]][[6]]

		sites_dfr = .merge_sites_and_muts(gr_sites, gr_maf, trinuc_sites, is_site=TRUE)
		upstream_dfr = .merge_sites_and_muts(gr_sites_upstream, gr_maf, trinuc_upstream, is_site=FALSE)
		downstream_dfr = .merge_sites_and_muts(gr_sites_downstream, gr_maf, trinuc_downstream, is_site=FALSE)
		dfr = rbind(upstream_dfr, sites_dfr, downstream_dfr)
		dfr$mut_rate = as.numeric(gsub("mutrate__", "", names(prepared_sites)[i]))
		dfr$mut_bin = i

		full_dfr = rbind(full_dfr, dfr)
	}
	full_dfr
}



#' Combines quadnucleotide counts and site information
#'
#' @param gr_sites GenomicRanges object of site/flanks chromosomes and positions
#' @param gr_maf GenomicRanges object with mcols containing quadnucleotide context or indel
#' @param trinuc_sites Data frame of counts for each trinucleotide counts extended to each quadnucleotide
#' @param is_site Boolean indicating whether mutation is within site or flanking region
#'
#' @return Data frame containing mutation and trinucleotide counts for each quadnucleotide context + indel. Includes boolean indicator of is_site
.merge_sites_and_muts = function(gr_sites, gr_maf, trinuc_sites, is_site) {
	muts_sites = .get_muts(gr_sites, gr_maf)
	dfr = cbind(is_site=is_site,
			merge(trinuc_sites, muts_sites, by.x="quadnuc", by.y="tag", all=T))
	dfr[is.na(dfr$muts_count), "muts_count"] = 0
	dfr
}



#' Counts number of mutations per quadnucleotide context
#'
#' @param gr_sites GenomicRanges object of site/flank chromosomes and positions
#' @param gr_maf GenomicRanges object with mcols containing quadnucleotide context or indel
#'
#' @return Data frame containing mutation count and each quadnucleotide context + indel
.get_muts = function(gr_sites, gr_maf) {

	# remove duplicate mutation calls
	gr_maf_in_sites = gr_maf[S4Vectors::subjectHits(GenomicRanges::findOverlaps(gr_sites, gr_maf))]
	gr_maf_in_sites = gr_maf_in_sites[!duplicated(GenomicRanges::mcols(gr_maf_in_sites)[,"ID"])]

	muts_sites = GenomicRanges::mcols(gr_maf_in_sites)[,"trinuc"]

	muts_sites = data.frame(muts_count = c(table(muts_sites)), stringsAsFactors=F)
	muts_sites$tag = rownames(muts_sites)
	muts_sites
}
